### PR TITLE
Breaking up the Rational Factors

### LINKED ISSUE #16296: collect_const() does cannot collect rationals
`collect_const()` does not seem to support the collecting of numeric rationals in SymPy 1.3. This is probably a duplicate of [13107](https://github.com/sympy/sympy/issues/13107), but I've provided a description below.

    In [41]: var('a:d')
    In [42]: f = a + b + c / 2 + d / 2
    In [43]: print(collect_const(f, Rational(1, 2), Numbers=True))
    a + b + (c/2 + d/2)

I'm expecting `a + b + 1 / 2 * (c + d)`

### PR DESCRIPTION
**Pull request by AnimeshSinha1309**
`Factors()` splits rational factors into its numerator and denominator.
`collect_const()` can now collect the common dividing factors in an expression.
 
#### References to other Issues or PRs
Fixes #16296 

#### Brief description of what is fixed or changed
Breaks down the numerator and denominator of Rational factors in a multiplicative expression, fixing the denominator collection issue that collect_const() was having.

#### Release Notes

* core
  * sympy.core.exprtools.Factors() now splits each rational factor into a seaprate numerator and denominator

### CODE REVIEW

File: sympy/core/exprtools.py [hunk 1]
```diff
@@ -355,6 +354,12 @@ def __init__(self, factors=None):  # Factors
             for _ in range(i):
                 c.remove(I)
             factors = dict(Mul._from_args(c).as_powers_dict())
+            # Handle all rational Coefficients
+            for f in list(factors.keys()):
+                if type(f) is Rational or type(f) is Half:
+                    factors[f.p] = (factors[f.p] if f.p in factors else 0) + factors[f]
```
Comment (jksuom) [inline sympy/core/exprtools.py]: Perhaps `factors.set_default(f.p, 0)` could be used.

File: sympy/core/exprtools.py [hunk 2]
```diff
@@ -355,6 +354,12 @@ def __init__(self, factors=None):  # Factors
             for _ in range(i):
                 c.remove(I)
             factors = dict(Mul._from_args(c).as_powers_dict())
+            # Handle all rational Coefficients
+            for f in list(factors.keys()):
```
Comment (jksuom) [inline sympy/core/exprtools.py]: Would it suffice to write `for f in factors.keys():`?

Comment (AnimeshSinha1309) [reply sympy/core/exprtools.py]: No, the `factors.keys()` method returns an iterator and as I change the dictionary the iterator throws an error.

File: sympy/core/exprtools.py [hunk 3]
```diff
@@ -355,6 +354,12 @@ def __init__(self, factors=None):  # Factors
...
+                if type(f) is Rational or type(f) is Half:
```
Comment (oscarbenjamin) [inline sympy/core/exprtools.py]: This could be `if isinstance(f, Rational)`

Comment (oscarbenjamin) [reply sympy/core/exprtools.py]: And then `Half won't need importing above.

**Comment by oscarbenjamin** [comment]
The tests seem to be breaking right now so I guess something here isn't working right.

This should have some new tests when it is working.

File: sympy/core/exprtools.py (hunk 1 above)
Comment (jksuom) [reply sympy/core/exprtools.py]: `f.p` is a Python `int`. It should be converted into a SymPy `Integer`.

File: sympy/core/exprtools.py (hunk 3 above)
Comment (AnimeshSinha1309) [reply sympy/core/exprtools.py]: Did both of those things.

File: sympy/core/exprtools.py [hunk 4]
```diff
@@ -356,9 +356,10 @@ def __init__(self, factors=None):  # Factors
...
             # Handle all rational Coefficients
             for f in list(factors.keys()):
-                if type(f) is Rational or type(f) is Half:
-                    factors[f.p] = (factors[f.p] if f.p in factors else 0) + factors[f]
-                    factors[f.q] = (factors[f.q] if f.q in factors else 0) - factors[f]
+                if isinstance(f, Rational):
+                    p, q = Integer(f.p), Integer(f.q)
+                    factors[p] = (factors[p] if p in factors else 0) + factors[f]
+                    factors[q] = (factors[q] if q in factors else 0) - factors[f]
                     factors.pop(f)
```
Comment (jksuom) [inline sympy/core/exprtools.py]: If `f` is integer, then `p` is `f` (and `q` is 1, which is not useful). This statement will delete `factors[p]`. The condition should probably be `isinstance(f, Rational) and not isinstance(f, Integer)`.

Comment (AnimeshSinha1309) [reply sympy/core/exprtools.py]: Thank you so much, got everything fixed and working.

**Comment by jksuom** [comment]
Please add a unit test.

File: sympy/simplify/tests/test_radsimp.py [hunk 5]
```diff
@@ -342,6 +342,9 @@ def test_collect_const():
     assert collect_sqrt(eq + 2) == \
         2*sqrt(sqrt(2) + 3)*(sqrt(5)*x + y) + 2
 
+    # issue 16296
+    assert collect_const(a + b + x/2 + y/2) != a + b + (x/2 + y/2)
```
Comment (jksuom) [inline sympy/simplify/tests/test_radsimp.py]: Perhaps this could be `== a + b + Mul(S.Half, x + y, evaluate=False)` (though it does not look so nice).

Comment (AnimeshSinha1309) [reply sympy/simplify/tests/test_radsimp.py]: Yeah, I did not know how to stop evaluation, thanks.

**Comment by AnimeshSinha1309** [comment]
Hope everything works well now, thanks a lot.

File: sympy/core/exprtools.py [hunk 6]
```diff
@@ -9,7 +9,7 @@
 from sympy.core.basic import Basic, preorder_traversal
 from sympy.core.expr import Expr
 from sympy.core.sympify import sympify
-from sympy.core.numbers import Rational, Integer, Number, I
+from sympy.core.numbers import Rational, Integer, Number, I, Half
```
Comment (jksuom) [inline sympy/core/exprtools.py]: The class `Half` is no more needed.

Comment (AnimeshSinha1309) [reply sympy/core/exprtools.py]: Done with it.

**Comment by jksuom** [comment]
Thanks, looks good.

**Pull request merged**